panel/widgets/chat.py

```diff
@@ -0,0 +1,1656 @@
+    show_reaction_icons: bool
+        Whether to display the reaction icons.
+    """
+    _ignored_refs: ClassVar[Tuple[str,...]] = ('value',)
+
+    value = param.Parameter(doc="""
+        The message contents. Can be any Python object that panel can display.""")
+
+    renderers = param.HookList(doc="""
+        A callable or list of callables that accept the value and return a
+        Panel object to render the value. If a list is provided, will
+        attempt to use the first renderer that does not raise an
+        exception. If None, will attempt to infer the renderer
+        from the value.""")
+
+    user = param.Parameter(default="User", doc="""
+        Name of the user who sent the message.""")
+
+    avatar = param.ClassSelector(default="", class_=(str, BinaryIO, ImageBase), doc="""
+        The avatar to use for the user. Can be a single character text, an emoji,
+        or anything supported by `pn.pane.Image`. If not set, checks if
+        the user is available in the default_avatars mapping; else uses the
+        first character of the name.""")
+
+    default_avatars = param.Dict(default=DEFAULT_AVATARS, doc="""
+        A default mapping of user names to their corresponding avatars
+        to use when the user is specified but the avatar is. You can modify, but not replace the
+        dictionary.""")
+
+    avatar_lookup = param.Callable(default=None, doc="""
+        A function that can lookup an `avatar` from a user name. The function signature should be
+        `(user: str) -> Avatar`. If this is set, `default_avatars` is disregarded.""")
+
+    reactions = param.List(doc="""
+        Reactions to associate with the message.""")
+
+    reaction_icons = param.ClassSelector(class_=(ChatReactionIcons, dict), doc="""
+        A mapping of reactions to their reaction icons; if not provided
+        defaults to `{"favorite": "heart"}`.""")
+
+    timestamp = param.Date(doc="""
+        Timestamp of the message. Defaults to the creation time.""")
+
+    timestamp_format = param.String(default="%H:%M", doc="The timestamp format.")
+
+    show_avatar = param.Boolean(default=True, doc="Whether to display the avatar of the user.")
+
+    show_user = param.Boolean(default=True, doc="Whether to display the name of the user.")
+
+    show_timestamp = param.Boolean(default=True, doc="Whether to display the timestamp of the message.")
+
+    show_reaction_icons = param.Boolean(default=True, doc="Whether to display the reaction icons.")
+
+    css_classes = param.List(default=["chat-entry"], doc="""
+        The CSS classes to apply to the widget.""")
+
+    _value_panel = param.Parameter(doc="The rendered value panel.")
+
+    _stylesheets: ClassVar[List[str]] = [
+        f"{CDN_DIST}css/chat_entry.css"
+    ]
+
+    _exit_stack: ClassVar[ExitStack | None] = None
+
+    def __init__(self, **params):
+        from ..param import ParamMethod  # circular imports
+
+        self._exit_stack = ExitStack()
```

Comment:
Every instance overrides the class variable? Is this desired?

Reply:
You're right; I actually think it should not be a class variable in the first place.

It should clear any instance level variables on cleanup:
```python
    def _cleanup(self, root=None) -> None:
        """
        Cleanup the exit stack.
        """
        if self._exit_stack is not None:
            self._exit_stack.close()
            self._exit_stack = None
        super()._cleanup()

```